### PR TITLE
*: let grpcproxy rise an error when its cert has non empty CN

### PR DESCRIPTION
**Pull request by mitake**
The change in this PR lets grpc-proxy deny to start when its cert has non empty CN. Because grpc-proxy terminates TLS connections, CN of its cert will be shared by its clients.

Fix https://github.com/etcd-io/etcd/issues/9521

### CODE REVIEW

File: embed/config.go [hunk 1]
```diff
@@ -315,6 +315,9 @@ type Config struct {
 	loggerCore        zapcore.Core
 	loggerWriteSyncer zapcore.WriteSyncer
 
+	// If EnableGRPCGateway is true, grpc gateway will be disabled.
+	EnableGRPCGateway bool `json:"enable-grpc-gateway"`
+
 	// TO BE DEPRECATED
```
[the remark this answers is not in the document]
Comment (mitake) [reply embed/config.go]: thanks, fixed

**Comment by mitake** [comment]
@gyuho could you take a look?

**Comment by mitake** [comment]
It seems that current failed tests are not related to this change.
- goword is generating a false error: https://travis-ci.com/etcd-io/etcd/jobs/169221991
- non deterministic runtime error in e2e is reported. I opened a new issue for this here: https://github.com/etcd-io/etcd/issues/10405

@gyuho could you take a look?

**Comment by mitake** [comment]
It seems that some wrong commits are not dropped from the branch... sorry for that. I'll fix it soon.

**Comment by mitake** [comment]
@gyuho @hexfusion rebased on the latest master, could you take a look?

File: pkg/transport/listener.go [hunk 2]
```diff
@@ -91,6 +91,9 @@ type TLSInfo struct {
 	// Logger logs TLS errors.
 	// If nil, all logs are discarded.
 	Logger *zap.Logger
+
+	// If EmptyCN is true, cert must not have non empty Common Name.
+	EmptyCN bool
 }
```
Comment (hexfusion) [inline pkg/transport/listener.go]: I think godoc will complain again needs to start with EmptyCN.

File: Documentation/op-guide/authentication.md [hunk 3]
```diff
@@ -167,7 +167,7 @@ $ etcdctl --user user --password password get foo
...
+As of version v3.2 if an etcd server is launched with the option `--client-cert-auth=true`, the field of Common Name (CN) in the client's TLS cert will be used as an etcd user. In this case, the common name authenticates the user and the client does not need a password. Note that if both of 1. `--client-cert-auth=true` is passed and CN is provided by the client, and 2. username and password are provided by the client, the username and password based authentication is prioritized. Note that this feature cannot be used with grpcproxy and gRPC-gateway. This is because grpcproxy terminates TLS from its client so all the clients share a cert of the proxy. gRPC-gateway uses a TLS connection internally for transforming HTTP request to gRPC request so it shares the same limitation. Therefore the clients cannot provide their CN to the server correctly. grpcproxy will cause an error and stop if a given cert has non empty CN. gRPC-proxy returns an error which indicates that the client has an non empty CN in its cert.
```
Comment (hexfusion) [inline Documentation/op-guide/authentication.md]: small nit s/grpcproxy/gRPC-proxy/g

File: pkg/transport/listener.go (hunk 2 above)
Comment (mitake) [reply pkg/transport/listener.go]: ah this problem again... sorry for that, will fix soon

File: Documentation/op-guide/authentication.md (hunk 3 above)
Comment (mitake) [reply Documentation/op-guide/authentication.md]: got it, thanks

**Comment by mitake** [comment]
@hexfusion @gyuho thanks! I'm merging this because the failed tests seem to be non deterministic (see also: https://github.com/etcd-io/etcd/issues/10405)

**Comment by jingyih** [comment]
cc @wenjiaswe

**Pull request merged**